File: informatics_custom_apps/ripl_customized_apps/report/average_indent_to_order_time_to_purchase_time___mgt/average_indent_to_order_time_to_purchase_time___mgt.py
# Copyright (c) 2026, Monil Kamboj and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import cint


def normalize_multiselect(value):
    if not value:
        return []
    if isinstance(value, str):
        return [v.strip() for v in value.split(",") if v.strip()]
    return value



def make_in_clause(field, values, values_dict):
    if not values:
        return ""

    placeholders = []

    for i, v in enumerate(values):
        key = f"{field.replace('.', '_')}_{i}"
        placeholders.append(f"%({key})s")
        values_dict[key] = v

    return f"AND {field} IN ({', '.join(placeholders)})"



def execute(filters=None):

    filters = filters or {}

    if filters.get("report_view") == "Detail":
        return get_detail_columns(), get_detail_data(filters)

    return get_summary_columns(), get_summary_data(filters)



def get_summary_columns():
    return [
        {"label": "Plant", "fieldname": "branch", "fieldtype": "Data", "width": 180},
        {"label": "Segment", "fieldname": "segment", "fieldtype": "Data", "width": 150},
        {"label": "Avg Buyer Delay(MR→PO)", "fieldname": "buyer_delay", "fieldtype": "Float", "width": 200},
        {"label": "Avg Supplier Delay(PO→GRN)", "fieldname": "supplier_delay", "fieldtype": "Float", "width": 250},
        {"label": "Avg Procurement Days", "fieldname": "total_procurement_days", "fieldtype": "Float", "width": 180},
        {"label": "Avg Delivery Delay", "fieldname": "delivery_delay", "fieldtype": "Float", "width": 180},
        {"label": "Total MRs", "fieldname": "total_mrs", "fieldtype": "Int", "width": 120},
        {"label": "Total POs", "fieldname": "total_pos", "fieldtype": "Int", "width": 120},
    ]


def get_summary_data(filters):

	values = {}

	companies = normalize_multiselect(filters.get("company"))
	branches = normalize_multiselect(filters.get("branch"))

	company_condition = make_in_clause("mr.company", companies, values)
	branch_condition = make_in_clause("mr.branch", branches, values)
	from_date = filters.get("from_date")
	to_date = filters.get("to_date")

	date_condition = ""

	if from_date and to_date:
		date_condition = " AND mr.transaction_date BETWEEN %(from_date)s AND %(to_date)s "
	elif from_date:
		date_condition = " AND mr.transaction_date >= %(from_date)s "
	elif to_date:
		date_condition = " AND mr.transaction_date <= %(to_date)s "

	values["from_date"] = from_date
	values["to_date"] = to_date

	query = f"""
		SELECT
			mr.branch AS branch,
			mri.segment AS segment,

			ROUND(AVG(DATEDIFF(po.transaction_date, mr.transaction_date)), 2) AS buyer_delay,
			ROUND(AVG(DATEDIFF(pr.posting_date, po.transaction_date)), 2) AS supplier_delay,
			ROUND(AVG(DATEDIFF(pr.posting_date, mr.transaction_date)), 2) AS total_procurement_days,
			ROUND(AVG(DATEDIFF(pr.posting_date, poi.schedule_date)), 2) AS delivery_delay,

			COUNT(DISTINCT mr.name) AS total_mrs,
			COUNT(DISTINCT po.name) AS total_pos,

			1 AS is_summary

		FROM `tabMaterial Request` mr

		LEFT JOIN `tabMaterial Request Item` mri
			ON mri.parent = mr.name

		LEFT JOIN `tabItem` i
			ON i.name = mri.item_code

		LEFT JOIN `tabPurchase Order Item` poi
			ON poi.material_request = mr.name
			AND poi.material_request_item = mri.name

		LEFT JOIN `tabPurchase Order` po
			ON po.name = poi.parent
			AND po.docstatus = 1

		LEFT JOIN `tabPurchase Receipt Item` pri
			ON pri.purchase_order_item = poi.name

		LEFT JOIN `tabPurchase Receipt` pr
			ON pr.name = pri.parent
			AND pr.docstatus = 1
			AND IFNULL(pr.is_return, 0) = 0

		WHERE
			mr.docstatus = 1
			AND mr.material_request_type = 'Purchase'
			AND LEFT(i.item_group, 6) NOT IN ('020201','020202','020203')
			{company_condition}
			{branch_condition}
			{date_condition}

		GROUP BY
			mr.branch,
			mri.segment

		ORDER BY
			AVG(DATEDIFF(pr.posting_date, mr.transaction_date)) DESC
	"""

	return frappe.db.sql(query, values, as_dict=True)



def get_detail_data(filters):

	values = {}

	companies = normalize_multiselect(filters.get("company"))
	branches = normalize_multiselect(filters.get("branch"))

	company_condition = make_in_clause("mr.company", companies, values)
	branch_condition = make_in_clause("mr.branch", branches, values)
	from_date = filters.get("from_date")
	to_date = filters.get("to_date")

	date_condition = ""

	if from_date and to_date:
		date_condition = " AND mr.transaction_date BETWEEN %(from_date)s AND %(to_date)s "
	elif from_date:
		date_condition = " AND mr.transaction_date >= %(from_date)s "
	elif to_date:
		date_condition = " AND mr.transaction_date <= %(to_date)s "

	values["from_date"] = from_date
	values["to_date"] = to_date

	query = f"""
		SELECT
			mr.name AS material_request,
			mr.transaction_date AS mr_date,

			mri.item_code,
			mri.item_name,
			i.item_group,
			i.stock_uom AS uom,

			mr.branch,
			mri.segment,

			ROUND(mri.qty,2) AS mr_qty,

			po.name AS purchase_order,
			po.transaction_date AS po_date,

			po.supplier,
			po.supplier_name,

			ROUND(poi.qty,2) AS po_qty,
			poi.schedule_date AS required_date,

			pr.name AS purchase_receipt,
			pr.posting_date AS receipt_date,

			ROUND(IFNULL(SUM(pri.qty),0),2) AS received_qty,

			DATEDIFF(po.transaction_date, mr.transaction_date) AS buyer_delay,
			DATEDIFF(pr.posting_date, po.transaction_date) AS supplier_delay,
			DATEDIFF(pr.posting_date, mr.transaction_date) AS total_procurement_days,
			DATEDIFF(pr.posting_date, poi.schedule_date) AS delivery_delay,

			CASE
				WHEN po.docstatus = 0 THEN 'PO Draft'
				WHEN po.docstatus = 1 AND IFNULL(SUM(pri.qty),0) = 0 THEN 'PO Submitted - No GRN'
				WHEN po.docstatus = 1 AND SUM(pri.qty) < poi.qty THEN 'Partial GRN'
				WHEN po.docstatus = 1 AND SUM(pri.qty) >= poi.qty THEN 'Full GRN'
				ELSE 'Unknown'
			END AS status

		FROM `tabMaterial Request` mr

		LEFT JOIN `tabMaterial Request Item` mri
			ON mri.parent = mr.name

		LEFT JOIN `tabItem` i
			ON i.name = mri.item_code

		LEFT JOIN `tabPurchase Order Item` poi
			ON poi.material_request = mr.name
			AND poi.material_request_item = mri.name

		LEFT JOIN `tabPurchase Order` po
			ON po.name = poi.parent

		LEFT JOIN `tabPurchase Receipt Item` pri
			ON pri.purchase_order_item = poi.name

		LEFT JOIN `tabPurchase Receipt` pr
			ON pr.name = pri.parent
			AND pr.docstatus = 1
			AND IFNULL(pr.is_return,0) = 0

		WHERE
			mr.docstatus = 1
			AND mr.material_request_type = 'Purchase'
			AND LEFT(i.item_group,6) NOT IN ('020201','020202','020203')
			{company_condition}
			{branch_condition}
			{date_condition}

		GROUP BY
			mr.name,
			mri.name,
			poi.name

		ORDER BY mr.transaction_date DESC
	"""

	return frappe.db.sql(query, values, as_dict=1)


def get_detail_columns():
    return [
        {"label": "Material Request", "fieldname": "material_request", "fieldtype": "Link", "options": "Material Request", "width": 190},
		{"label": "Item Code", "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 340,"align": "left"},
        {"label": "Plant", "fieldname": "branch", "fieldtype": "Link", "options": "Branch", "width": 140},
        {"label": "Segment", "fieldname": "segment", "fieldtype": "Data", "width": 120},
        {"label": "Buyer Delay(MR->PO days)", "fieldname": "buyer_delay", "fieldtype": "Int", "width": 190},
        {"label": "Supplier Delay(PO->GRN days)", "fieldname": "supplier_delay", "fieldtype": "Int", "width": 210},
        {"label": "Total Procurement Days", "fieldname": "total_procurement_days", "fieldtype": "Int", "width": 190},
        {"label": "Delivery Delay", "fieldname": "delivery_delay", "fieldtype": "Int", "width": 140},
        {"label": "MR Date", "fieldname": "mr_date", "fieldtype": "Date", "width": 110},
        {"label": "PO Date", "fieldname": "po_date", "fieldtype": "Date", "width": 110},
        {"label": "Required Date", "fieldname": "required_date", "fieldtype": "Date", "width": 120},
        {"label": "Receipt Date", "fieldname": "receipt_date", "fieldtype": "Date", "width": 120},
        {"label": "PO / GRN Status", "fieldname": "status", "fieldtype": "Data", "width": 180},
		{"label": "Item Name", "fieldname": "item_name", "fieldtype": "Data", "width": 200},
		{"label": "Item Group", "fieldname": "item_group", "fieldtype": "Link", "options": "Item Group", "width": 150},
		{"label": "UOM", "fieldname": "uom", "fieldtype": "Data", "width": 80},
		{"label": "MR Qty", "fieldname": "mr_qty", "fieldtype": "Float", "width": 120},
		{"label": "Purchase Order", "fieldname": "purchase_order", "fieldtype": "Link", "options": "Purchase Order", "width": 170},
		{"label": "Supplier Code", "fieldname": "supplier", "fieldtype": "Link", "options": "Supplier", "width": 150},
		{"label": "Supplier Name", "fieldname": "supplier_name", "fieldtype": "Data", "width": 200},
		{"label": "PO Qty", "fieldname": "po_qty", "fieldtype": "Float", "width": 120},
		{"label": "Purchase Receipt", "fieldname": "purchase_receipt", "fieldtype": "Link", "options": "Purchase Receipt", "width": 170},
		{"label": "Received Qty", "fieldname": "received_qty", "fieldtype": "Float", "width": 120},
         
    ]

File: informatics_custom_apps/ripl_customized_apps/report/average_indent_to_order_time_to_purchase_time___mgt/average_indent_to_order_time_to_purchase_time___mgt.js
frappe.query_reports["Average Indent To Order Time To Purchase Time - MGT"] = {

    onload: function (report) {

        if (report.__bound_back_to_summary) return;
        report.__bound_back_to_summary = true;

        report.page.add_inner_button(__("Back To Summary"), async function () {

            const from_date = frappe.query_report.get_filter_value("from_date");
            const to_date = frappe.query_report.get_filter_value("to_date");

            await frappe.query_report.set_filter_value("show_detail", 0);
            await frappe.query_report.set_filter_value("branch", null);
            await frappe.query_report.set_filter_value("segment", null);
            await frappe.query_report.set_filter_value("drill_metric", null);
            await frappe.query_report.set_filter_value("report_view", "Summary");

            await frappe.query_report.set_filter_value("from_date", from_date);
            await frappe.query_report.set_filter_value("to_date", to_date);

            frappe.query_report.refresh();
        });

        frappe.call({
            method: "erpnext.accounts.utils.get_fiscal_year",
            args: {
                date: frappe.datetime.get_today(),
                company: frappe.defaults.get_user_default("Company")
            },
            callback: function (r) {

                if (!r.message) return;

                if (!frappe.query_report.get_filter_value("from_date")) {
                    frappe.query_report.set_filter_value("from_date", r.message[1]);
                }

                if (!frappe.query_report.get_filter_value("to_date")) {
                    frappe.query_report.set_filter_value("to_date", frappe.datetime.get_today());
                }
            }
        });
    },

    formatter: function (value, row, column, data, default_formatter) {

        value = default_formatter(value, row, column, data);

        const drill_columns = [
            "buyer_delay",
            "supplier_delay",
            "total_procurement_days",
            "delivery_delay"
        ];

        if (data && data.is_summary && drill_columns.includes(column.fieldname)) {

            value = `<a href="javascript:void(0)"
                onclick="show_procurement_detail(
                    '${data.branch || ""}',
                    '${data.segment || ""}',
                    '${column.fieldname}'
                ); return false;">
                ${value}
            </a>`;
        }

        return value;
    },

    filters: [

        {
            fieldname: "company",
            label: "Company",
            fieldtype: "MultiSelectList",
            get_data: function (txt) {
                return frappe.db.get_link_options("Company", txt);
            }
        },

        {
            fieldname: "branch",
            label: "Plant",
            fieldtype: "Link",
            options: "Branch",
            hidden: 1
        },

        {
            fieldname: "from_date",
            label: "From Date",
            fieldtype: "Date",
            reqd: 1
        },

        {
            fieldname: "to_date",
            label: "To Date",
            fieldtype: "Date",
            reqd: 1
        },

        {
            fieldname: "segment",
            label: "Segment",
            fieldtype: "Data",
            hidden: 1
        },

        {
            fieldname: "drill_metric",
            label: "Drill Metric",
            fieldtype: "Data",
            hidden: 1
        },

        {
            fieldname: "show_detail",
            label: "Show Detail",
            fieldtype: "Check",
            default: 0,
            hidden: 1
        },

     {
			fieldname: "report_view",
			label: "Report View",
			fieldtype: "Select",
			options: "Summary\nDetail",
			default: "Summary",

			on_change: function () {

				const view = frappe.query_report.get_filter_value("report_view");

				if (view === "Summary") {

					frappe.query_report.set_filter_value("show_detail", 0);
					frappe.query_report.set_filter_value("branch", null);
					frappe.query_report.set_filter_value("segment", null);
					frappe.query_report.set_filter_value("drill_metric", null);

					frappe.query_report.refresh();
				}

				if (view === "Detail") {

					frappe.query_report.set_filter_value("show_detail", 1);

					frappe.query_report.refresh();
				}
			}
		}
    ]
};

window.show_procurement_detail = async function (branch, segment, metric) {

    const from_date = frappe.query_report.get_filter_value("from_date");
    const to_date = frappe.query_report.get_filter_value("to_date");

    await frappe.query_report.set_filter_value("branch", branch);
    await frappe.query_report.set_filter_value("segment", segment);
    await frappe.query_report.set_filter_value("drill_metric", metric);
    await frappe.query_report.set_filter_value("show_detail", 1);
    await frappe.query_report.set_filter_value("report_view", "Detail");

    await frappe.query_report.set_filter_value("from_date", from_date);
    await frappe.query_report.set_filter_value("to_date", to_date);

    frappe.query_report.refresh();
};